### PR TITLE
GH-43267: [C#] Correctly import sliced arrays through the C Data interface

### LINKED ISSUE #43267: [C#] C Data Interface import computes incorrect buffer sizes when offset is non-zero
### Describe the bug, including details regarding any error messages, version, and platform.

Code to reproduce the problem as an XUnit test:
```c#
[Fact]
public unsafe void RoundTripArrayWithOffset()
{
    Int32Array array = new Int32Array.Builder()
        .AppendRange(new[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 })
        .Build();
    IArrowArray sliced = array.Slice(2, 6);
    CArrowArray* cArray = CArrowArray.Create();
    CArrowArrayExporter.ExportArray(sliced, cArray);
    using (var importedSlice = (Int32Array)CArrowArrayImporter.ImportArray(cArray, array.Data.DataType))
    {
        Assert.Equal(6, importedSlice.Length);  // OK
        Assert.Equal(2, importedSlice.Offset);  // OK
        Assert.Equal(2, importedSlice.GetValue(0));  // Throws
    }
    CArrowArray.Free(cArray);
}
```
This throws:
```
System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.

System.ArgumentOutOfRangeException
Specified argument was out of the range of valid values.
   at Apache.Arrow.PrimitiveArray`1.get_Values() in /home/adam/dev/arrow/csharp/src/Apache.Arrow/Arrays/PrimitiveArray.cs:line 34
   at Apache.Arrow.Tests.CDataInterfaceDataTests.RoundTripArrayWithOffset() in /home/adam/dev/arrow/csharp/test/Apache.Arrow.Tests/CDataInterfaceDataTests.cs:line 83
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
```

The problem is that when importing the buffers, we assume the buffer length is at least `array.length * typeByteWidth` (see [ImportFixedWidthBuffers](https://github.com/apache/arrow/blob/6845bb64e4e4da42b146ca5227db0a1c4ee33f94/csharp/src/Apache.Arrow/C/CArrowArrayImporter.cs#L393)), but really it should be `(array.length + array.offset) * typeByteWidth`. This example is for a fixed width array type, but other more complex array types appear to have the same problem.

I don't have any immediate plans to fix this myself, but noticed this problem as part of looking at #43266.

### Component(s)

C#

### PR DESCRIPTION
**Pull request by CurtHagenlocher**
### What changes are included in this PR?

Changes to the C Data importer to correctly handle nonzero offsets.

### Are these changes tested?

Yes

### Are there any user-facing changes?

No

Closes #43267 
* GitHub Issue: #43267

### CODE REVIEW

File: csharp/src/Apache.Arrow/C/CArrowArrayImporter.cs [hunk 1]
```diff
@@ -148,6 +148,11 @@ public RecordBatch GetAsRecordBatch(Schema schema)
             }
 
             private ArrayData GetAsArrayData(CArrowArray* cArray, IArrowType type)
+            {
```
Comment (CurtHagenlocher) [inline csharp/src/Apache.Arrow/C/CArrowArrayImporter.cs:151]: Derp; forgot to revert this

**Comment by CurtHagenlocher** [comment]
@adamreeve, I would appreciate a quick look at these changes.

File: csharp/src/Apache.Arrow/Apache.Arrow.csproj [hunk 2]
```diff
@@ -7,18 +7,16 @@
     <Description>Apache Arrow is a cross-language development platform for in-memory data. It specifies a standardized language-independent columnar memory format for flat and hierarchical data, organized for efficient analytic operations on modern hardware.</Description>
   </PropertyGroup>
 
-  <PropertyGroup Condition="'$(IsWindows)'=='true'">
+  <PropertyGroup>
```
Comment (adamreeve) [inline csharp/src/Apache.Arrow/Apache.Arrow.csproj:10]: Just checking this was intentionally included in the PR? Seems reasonable to now build for net462 on non-Windows too, this built fine for me on Linux and I guess something made sure the reference assemblies / targeting pack was available.

**Review by adamreeve — APPROVED** [review]
Looks good to me thanks Curt :+1:

File: csharp/src/Apache.Arrow/Apache.Arrow.csproj (hunk 2 above)
Comment (CurtHagenlocher) [reply csharp/src/Apache.Arrow/Apache.Arrow.csproj:10]: It was intentional, though arguably I should have put it into a separate PR. What I didn't realize when adding net462 support back was that the official NuGet packages aren't built on Windows and so they were missing the 4.6.2 assembly.

**Pull request merged**